DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
-    "    tower_labels, \\\n",
-    "    tower_gradients, \\\n",
-    "    tower_distances, \\\n",
-    "    tower_total_losses, \\\n",
-    "    avg_loss, \\\n",
-    "    avg_accuracy \\\n",
-    "    = get_tower_results(data_sets.train, optimizer)\n",
+    "    with graph.as_default():\n",
+    "        if model_path is None:\n",
+    "            # Init all variables for first use\n",
+    "            session.run(tf.initialize_all_variables())\n",
+    "        else:\n",
+    "            # Loading the model into the session\n",
+    "            execution_context[3].restore(session, model_path)\n",
+    "\n",
+    "        # Start queue runner threads\n",
+    "        runner_threads = tf.train.start_queue_runners(sess=session)\n",
     "\n",
-    "    # Average tower gradients\n",
-    "    avg_tower_gradients = average_gradients(tower_gradients)\n",
+    "        # Start importer's queue threads\n",
+    "        queue_threads = execution_context[1].start_queue_threads(session)\n",
     "\n",
-    "    # Add logging of averaged gradients\n",
-    "    log_grads_and_vars(avg_tower_gradients)\n",
+    "    return session"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following helper method stops an execution contect. \n",
+    "Before closing the provided `session`, it will persist the contained model to disk. \n",
+    "The model's filename is constructed from `checkpoint_path` and `global_step` and will be returned."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def stop_execution_context(execution_context, session, checkpoint_path=None, global_step=None):\n",
+    "    \n",
+    "    # If the model is not persisted, we'll return 'None'\n",
+    "    hibernation_path = None\n",
+    "    \n",
+    "    if checkpoint_path is not None and global_step is not None:\n",
+    "        # Saving session's model into checkpoint dir\n",
+    "        hibernation_path = execution_context[3].save(session, checkpoint_path, global_step=global_step)\n",
+    "        \n",
+    "    # Free all allocated resources of the session\n",
+    "    session.close()\n",
+    "    \n",
+    "    return hibernation_path"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's introduce the main routine for training and inference. \n",
+    "It takes a started execution context (given by `execution_context` and `session`), an optional `epoch` index and a flag wheather to calculate WER report data (`query_report`) or not.\n",
+    "Its main duty is to iterate over all batches and calculate the mean loss.\n",
+    "If an epoch is provided, it will also do back propagation - aka training.\n",
+    "It will return a batch set result tuple which consists of the mean loss and - if asked for report data - the mean accuracy and individual results tuple."
```

Comment:
I'm having trouble scanning this sentence, definitely needs some commas or possible rephrasing... Rather than having a variable return value, could the `query_report` part just be a separate function?

The function is a bit hard to follow because of all the conditional paths right now, it'd be good if we could decouple it somewhat, or at least rearrange it so that the paths don't have to be interleaved quite so much.